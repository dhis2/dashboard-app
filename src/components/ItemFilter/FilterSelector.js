import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import Popover from '@material-ui/core/Popover';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import { withTheme } from '@material-ui/core/styles';

import i18n from '@dhis2/d2-i18n';
import { DimensionsPanel } from '@dhis2/d2-ui-analytics';

import { Button } from '@dhis2/ui-core';
import FilterDialog from './FilterDialog';

import { sGetSettingsDisplayNameProperty } from '../../reducers/settings';
import { sGetActiveModalDimension } from '../../reducers/activeModalDimension';
import { sGetDimensions } from '../../reducers/dimensions';
import { sGetFiltersKeys } from '../../reducers/itemFilters';
import { sGetEditItemFiltersRoot } from '../../reducers/editItemFilters';
import { acAddItemFilter, acRemoveItemFilter } from '../../actions/itemFilters';
import {
    acRemoveEditItemFilter,
    acSetEditItemFilters,
} from '../../actions/editItemFilters';
import {
    acClearActiveModalDimension,
    acSetActiveModalDimension,
} from '../../actions/activeModalDimension';

class FilterSelector extends Component {
    constructor(props) {
        super(props);

        this.state = {
            anchorEl: null,
        };
    }

    openPanel = event => {
        this.setState({ anchorEl: event.currentTarget });
    };

    closePanel = () => {
        this.setState({ anchorEl: null });
    };

    closeDialog = () => {
        this.setState({ anchorEl: null });

        this.props.clearActiveModalDimension();
    };

    selectDimension = id => {
        this.props.setActiveModalDimension({
            id,
            name: this.props.dimensions[id].name,
        });
    };

    onSelectItems = ({ dimensionId, items }) => {
        this.props.setEditItemFilters({
            ...this.props.selectedItems,
            [dimensionId]: items,
        });
    };

    onDeselectItems = ({ dimensionId, itemIdsToRemove }) => {
        const oldList = this.props.selectedItems[dimensionId] || [];
        const newList = oldList.filter(
            item => !itemIdsToRemove.includes(item.id)
        );

        if (newList.length) {
            this.props.setEditItemFilters({
                ...this.props.selectedItems,
                [dimensionId]: newList,
            });
        } else {
            this.props.removeEditItemFilter(dimensionId);
        }
    };

    onReorderItems = ({ dimensionId, itemIds }) => {
        const oldList = this.props.selectedItems[dimensionId] || [];
        const reorderedList = itemIds.map(id =>
            oldList.find(item => item.id === id)
        );

        this.props.setEditItemFilters({
            ...this.props.selectedItems,
            [dimensionId]: reorderedList,
        });
    };

    saveFilter = id => {
        const filterItems = this.props.selectedItems[id];

        if (filterItems && filterItems.length) {
            this.props.addItemFilter({
                id,
                value: [...filterItems],
            });
        } else {
            this.props.removeItemFilter(id);
        }

        this.closeDialog();
    };

    render() {
        const {
            theme,
            displayNameProperty,
            dimension,
            dimensions,
            selectedDimensions,
            selectedItems,
        } = this.props;

        return (
            <Fragment>
                <Button onClick={this.openPanel}>
                    {i18n.t('Add filter')}
                    <ArrowDropDownIcon />
                </Button>
                <Popover
                    open={Boolean(this.state.anchorEl)}
                    anchorEl={this.state.anchorEl}
                    anchorOrigin={{ vertical: 'bottom', horizontal: 'left' }}
                    onClose={this.closePanel}
                    style={{
                        height: '100%',
                        fontFamily: theme.typography.fontFamily,
                    }}
                >
                    <DimensionsPanel
                        dimensions={dimensions}
                        onDimensionClick={this.selectDimension}
                        selectedIds={selectedDimensions}
                    />
                </Popover>
                {dimension ? (
                    <FilterDialog
                        displayNameProperty={displayNameProperty}
                        dimension={dimension}
                        selectedItems={selectedItems[dimension.id] || []}
                        onSelect={this.onSelectItems}
                        onDeselect={this.onDeselectItems}
                        onReorder={this.onReorderItems}
                        onClose={this.closeDialog}
                        onConfirm={this.saveFilter}
                    />
                ) : null}
            </Fragment>
        );
    }
}

const mapStateToProps = state => ({
    displayNameProperty: sGetSettingsDisplayNameProperty(state),
    dimension: sGetActiveModalDimension(state),
    dimensions: sGetDimensions(state),
    selectedDimensions: sGetFiltersKeys(state),
    selectedItems: sGetEditItemFiltersRoot(state),
});

export default connect(
    mapStateToProps,
    {
        clearActiveModalDimension: acClearActiveModalDimension,
        setActiveModalDimension: acSetActiveModalDimension,
        addItemFilter: acAddItemFilter,
        removeItemFilter: acRemoveItemFilter,
        removeEditItemFilter: acRemoveEditItemFilter,
        setEditItemFilters: acSetEditItemFilters,
    }
)(withTheme()(FilterSelector));
