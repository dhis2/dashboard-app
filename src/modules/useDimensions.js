import { apiFetchDimensions } from '@dhis2/analytics'
import { useDataEngine } from '@dhis2/app-runtime'
import { useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { acSetDimensions } from '../actions/dimensions'
import { useUserSettings } from '../components/UserSettingsProvider'
import getFilteredDimensions from './getFilteredDimensions'

const useDimensions = doFetch => {
    const dataEngine = useDataEngine()
    const { userSettings } = useUserSettings()
    const dimensions = useSelector(state => state.dimensions)
    const dispatch = useDispatch()

    useEffect(() => {
        const fetchDimensions = async () => {
            try {
                const unfilteredDimensions = await apiFetchDimensions(
                    dataEngine,
                    userSettings.displayProperty
                )

                dispatch(
                    acSetDimensions(getFilteredDimensions(unfilteredDimensions))
                )
            } catch (e) {
                console.error(e)
            }
        }

        if (!dimensions.length && doFetch && userSettings.displayProperty) {
            fetchDimensions()
        }
    }, [
        dimensions,
        doFetch,
        userSettings.displayProperty,
        dataEngine,
        dispatch,
    ])

    return dimensions
}

export default useDimensions
