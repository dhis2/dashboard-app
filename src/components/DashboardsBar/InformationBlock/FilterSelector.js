import { DimensionsPanel } from '@dhis2/analytics'
import { useDhis2ConnectionStatus } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { Card, colors, IconFilter24 } from '@dhis2/ui'
import isEmpty from 'lodash/isEmpty.js'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { connect } from 'react-redux'
import {
    acClearActiveModalDimension,
    acSetActiveModalDimension,
} from '../../../actions/activeModalDimension.js'
import useDimensions from '../../../modules/useDimensions.js'
import { sGetActiveModalDimension } from '../../../reducers/activeModalDimension.js'
import { sGetItemFiltersRoot } from '../../../reducers/itemFilters.js'
import DropdownButton from '../../DropdownButton/DropdownButton.js'
import FilterDialog from './FilterDialog.js'

const FilterSelector = (props) => {
    const [filterDialogIsOpen, setFilterDialogIsOpen] = useState(false)
    const dimensions = useDimensions(filterDialogIsOpen)
    const { isDisconnected: offline } = useDhis2ConnectionStatus()

    const toggleFilterDialogIsOpen = () =>
        setFilterDialogIsOpen(!filterDialogIsOpen)

    const onCloseDialog = () => {
        setFilterDialogIsOpen(false)

        props.clearActiveModalDimension()
    }

    const selectDimension = (id) => {
        props.setActiveModalDimension(
            dimensions.find((dimension) => dimension.id === id)
        )
    }

    const filterDimensions = () => {
        if (!props.restrictFilters) {
            return dimensions
        } else {
            return dimensions.filter((d) =>
                [...props.allowedFilters].includes(d.id)
            )
        }
    }

    const getFilterSelector = () => (
        <Card dataTest="dashboard-filter-popover">
            <DimensionsPanel
                style={{ width: '320px' }}
                dimensions={filterDimensions()}
                onDimensionClick={selectDimension}
                selectedIds={Object.keys(props.initiallySelectedItems)}
            />
        </Card>
    )

    return props.restrictFilters && !props.allowedFilters?.length ? null : (
        <>
            <DropdownButton
                secondary
                small
                open={filterDialogIsOpen}
                disabled={offline}
                onClick={toggleFilterDialogIsOpen}
                icon={<IconFilter24 color={colors.grey700} />}
                component={getFilterSelector()}
            >
                {i18n.t('Filter')}
            </DropdownButton>
            {!isEmpty(props.dimension) ? (
                <FilterDialog
                    dimension={props.dimension}
                    onClose={onCloseDialog}
                />
            ) : null}
        </>
    )
}

const mapStateToProps = (state) => ({
    dimension: sGetActiveModalDimension(state),
    initiallySelectedItems: sGetItemFiltersRoot(state),
})

FilterSelector.propTypes = {
    allowedFilters: PropTypes.array,
    clearActiveModalDimension: PropTypes.func,
    dimension: PropTypes.object,
    initiallySelectedItems: PropTypes.object,
    restrictFilters: PropTypes.bool,
    setActiveModalDimension: PropTypes.func,
}

export default connect(mapStateToProps, {
    clearActiveModalDimension: acClearActiveModalDimension,
    setActiveModalDimension: acSetActiveModalDimension,
})(FilterSelector)
