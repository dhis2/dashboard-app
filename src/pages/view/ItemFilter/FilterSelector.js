import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import isEmpty from 'lodash/isEmpty'
import i18n from '@dhis2/d2-i18n'
import { DimensionsPanel } from '@dhis2/analytics'
import { Card, DropdownButton, colors, IconFilter24 } from '@dhis2/ui'
import FilterDialog from './FilterDialog'

import { sGetActiveModalDimension } from '../../../reducers/activeModalDimension'
import { sGetItemFiltersRoot } from '../../../reducers/itemFilters'
import {
    acClearActiveModalDimension,
    acSetActiveModalDimension,
} from '../../../actions/activeModalDimension'
import useDimensions from '../../../modules/useDimensions'

import classes from './styles/FilterSelector.module.css'

const FilterSelector = props => {
    const [showPopover, setShowPopover] = useState(false)
    const dimensions = useDimensions(showPopover)

    const onCloseDialog = () => {
        setShowPopover(false)

        props.clearActiveModalDimension()
    }

    const selectDimension = id => {
        props.setActiveModalDimension(
            dimensions.find(dimension => dimension.id === id)
        )
    }

    const filterDimensions = () => {
        if (!props.restrictFilters) {
            return dimensions
        } else {
            return dimensions.filter(d =>
                [...props.allowedFilters].includes(d.id)
            )
        }
    }

    const getFilterSelector = () => (
        <Card>
            <DimensionsPanel
                style={{ width: '320px' }}
                dimensions={filterDimensions()}
                onDimensionClick={selectDimension}
                selectedIds={Object.keys(props.initiallySelectedItems)}
            />
        </Card>
    )

    return props.restrictFilters && !props.allowedFilters?.length ? null : (
        <>
            <span className={classes.buttonContainer}>
                <DropdownButton
                    onClick={() => setShowPopover(true)}
                    icon={<IconFilter24 color={colors.grey700} />}
                    component={getFilterSelector()}
                >
                    {i18n.t('Add filter')}
                </DropdownButton>
            </span>
            {!isEmpty(props.dimension) ? (
                <FilterDialog
                    dimension={props.dimension}
                    onClose={onCloseDialog}
                />
            ) : null}
        </>
    )
}

const mapStateToProps = state => ({
    dimension: sGetActiveModalDimension(state),
    initiallySelectedItems: sGetItemFiltersRoot(state),
})

FilterSelector.propTypes = {
    allowedFilters: PropTypes.array,
    clearActiveModalDimension: PropTypes.func,
    dimension: PropTypes.object,
    initiallySelectedItems: PropTypes.object,
    restrictFilters: PropTypes.bool,
    setActiveModalDimension: PropTypes.func,
}

export default connect(mapStateToProps, {
    clearActiveModalDimension: acClearActiveModalDimension,
    setActiveModalDimension: acSetActiveModalDimension,
})(FilterSelector)
