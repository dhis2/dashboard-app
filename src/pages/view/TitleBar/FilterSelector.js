import { DimensionsPanel } from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'
import { Card, colors, IconFilter24 } from '@dhis2/ui'
import isEmpty from 'lodash/isEmpty'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { connect } from 'react-redux'
import {
    acClearActiveModalDimension,
    acSetActiveModalDimension,
} from '../../../actions/activeModalDimension'
import DropdownButton from '../../../components/DropdownButton/DropdownButton'
import useDimensions from '../../../modules/useDimensions'
import { sGetActiveModalDimension } from '../../../reducers/activeModalDimension'
import { sGetItemFiltersRoot } from '../../../reducers/itemFilters'
import FilterDialog from './FilterDialog'
import classes from './styles/FilterSelector.module.css'

const FilterSelector = props => {
    const [filterDialogIsOpen, setFilterDialogIsOpen] = useState(false)
    const dimensions = useDimensions(filterDialogIsOpen)

    const toggleFilterDialogIsOpen = () =>
        setFilterDialogIsOpen(!filterDialogIsOpen)

    const onCloseDialog = () => {
        setFilterDialogIsOpen(false)

        props.clearActiveModalDimension()
    }

    const selectDimension = id => {
        props.setActiveModalDimension(
            dimensions.find(dimension => dimension.id === id)
        )
    }

    const filterDimensions = () => {
        if (!props.restrictFilters) {
            return dimensions
        } else {
            return dimensions.filter(d =>
                [...props.allowedFilters].includes(d.id)
            )
        }
    }

    const getFilterSelector = () => (
        <Card dataTest="dashboard-filter-popover">
            <DimensionsPanel
                style={{ width: '320px' }}
                dimensions={filterDimensions()}
                onDimensionClick={selectDimension}
                selectedIds={Object.keys(props.initiallySelectedItems)}
            />
        </Card>
    )

    return props.restrictFilters && !props.allowedFilters?.length ? null : (
        <>
            <span className={classes.buttonContainer}>
                <DropdownButton
                    open={filterDialogIsOpen}
                    onClick={toggleFilterDialogIsOpen}
                    icon={<IconFilter24 color={colors.grey700} />}
                    component={getFilterSelector()}
                >
                    {i18n.t('Add filter')}
                </DropdownButton>
            </span>
            {!isEmpty(props.dimension) ? (
                <FilterDialog
                    dimension={props.dimension}
                    onClose={onCloseDialog}
                />
            ) : null}
        </>
    )
}

const mapStateToProps = state => ({
    dimension: sGetActiveModalDimension(state),
    initiallySelectedItems: sGetItemFiltersRoot(state),
})

FilterSelector.propTypes = {
    allowedFilters: PropTypes.array,
    clearActiveModalDimension: PropTypes.func,
    dimension: PropTypes.object,
    initiallySelectedItems: PropTypes.object,
    restrictFilters: PropTypes.bool,
    setActiveModalDimension: PropTypes.func,
}

export default connect(mapStateToProps, {
    clearActiveModalDimension: acClearActiveModalDimension,
    setActiveModalDimension: acSetActiveModalDimension,
})(FilterSelector)
