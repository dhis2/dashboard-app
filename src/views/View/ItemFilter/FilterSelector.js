import React, { useState, useRef } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown'
import isEmpty from 'lodash/isEmpty'
import i18n from '@dhis2/d2-i18n'
import { DimensionsPanel } from '@dhis2/analytics'
import { Button, Popover } from '@dhis2/ui'
import FilterDialog from './FilterDialog'

import { sGetActiveModalDimension } from '../../../reducers/activeModalDimension'
import { sGetItemFiltersRoot } from '../../../reducers/itemFilters'
import {
    acClearActiveModalDimension,
    acSetActiveModalDimension,
} from '../../../actions/activeModalDimension'
import useDimensions from '../../../modules/useDimensions'

import classes from './styles/FilterSelector.module.css'

const FilterSelector = props => {
    const [showPopover, setShowPopover] = useState(false)
    const dimensions = useDimensions(showPopover)

    const ref = useRef(null)

    const onCloseDialog = () => {
        setShowPopover(false)

        props.clearActiveModalDimension()
    }

    const selectDimension = id => {
        props.setActiveModalDimension(
            dimensions.find(dimension => dimension.id === id)
        )
    }

    const filterDimensions = () => {
        if (!props.restrictFilters) {
            return dimensions
        } else {
            return dimensions.filter(d =>
                [...props.allowedFilters].includes(d.id)
            )
        }
    }

    return props.restrictFilters && !props.allowedFilters?.length ? null : (
        <>
            <span className={classes.buttonContainer} ref={ref}>
                <Button onClick={() => setShowPopover(true)}>
                    {i18n.t('Add filter')}
                    <ArrowDropDownIcon />
                </Button>
            </span>
            {showPopover && (
                <Popover
                    onClickOutside={onCloseDialog}
                    reference={ref}
                    arrow={true}
                    placement="bottom-start"
                    dataTest="dashboard-filter-popover"
                >
                    <div className={classes.popover}>
                        <DimensionsPanel
                            style={{ width: '320px' }}
                            dimensions={filterDimensions()}
                            onDimensionClick={selectDimension}
                            selectedIds={Object.keys(
                                props.initiallySelectedItems
                            )}
                        />
                    </div>
                </Popover>
            )}
            {!isEmpty(props.dimension) ? (
                <FilterDialog
                    dimension={props.dimension}
                    onClose={onCloseDialog}
                />
            ) : null}
        </>
    )
}

const mapStateToProps = state => ({
    dimension: sGetActiveModalDimension(state),
    initiallySelectedItems: sGetItemFiltersRoot(state),
})

FilterSelector.propTypes = {
    allowedFilters: PropTypes.array,
    clearActiveModalDimension: PropTypes.func,
    dimension: PropTypes.object,
    initiallySelectedItems: PropTypes.object,
    restrictFilters: PropTypes.bool,
    setActiveModalDimension: PropTypes.func,
}

export default connect(mapStateToProps, {
    clearActiveModalDimension: acClearActiveModalDimension,
    setActiveModalDimension: acSetActiveModalDimension,
})(FilterSelector)
