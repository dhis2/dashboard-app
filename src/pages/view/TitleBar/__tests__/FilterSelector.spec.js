import { useDhis2ConnectionStatus } from '@dhis2/app-runtime'
import { render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import useDimensions from '../../../../modules/useDimensions'
import FilterSelector from '../FilterSelector'

const mockStore = configureMockStore()

jest.mock('@dhis2/app-runtime', () => ({
    useDhis2ConnectionStatus: jest.fn(() => ({ isDisconnected: false })),
}))

jest.mock('../../../../modules/useDimensions', () => jest.fn())
useDimensions.mockImplementation(() => ['Moomin', 'Snorkmaiden'])

test('is disabled when offline', () => {
    useDhis2ConnectionStatus.mockImplementationOnce(
        jest.fn(() => ({ isDisconnected: true }))
    )

    const store = { activeModalDimension: {}, itemFilters: {} }

    const props = {
        allowedFilters: [],
        restrictFilters: false,
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <FilterSelector {...props} />
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('is enabled when online', () => {
    // useDhis2ConnectionStatus.mockImplementation(jest.fn(() => ({ isDisconnected: false })))

    const store = { activeModalDimension: {}, itemFilters: {} }

    const props = {
        allowedFilters: [],
        restrictFilters: false,
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <FilterSelector {...props} />
        </Provider>
    )
    expect(container).toMatchSnapshot()
})

test('is null when no filters are restricted and no filters are allowed', () => {
    const store = { activeModalDimension: {}, itemFilters: {} }

    const props = {
        allowedFilters: [],
        restrictFilters: true,
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <FilterSelector {...props} />
        </Provider>
    )
    expect(container.firstChild).toBeNull()
})

test('is null when no filters are restricted and allowedFilters undefined', () => {
    const store = { activeModalDimension: {}, itemFilters: {} }

    const props = {
        restrictFilters: true,
    }

    const { container } = render(
        <Provider store={mockStore(store)}>
            <FilterSelector {...props} />
        </Provider>
    )
    expect(container.firstChild).toBeNull()
})

test('shows button when filters are restricted and at least one filter is allowed', () => {
    const store = { activeModalDimension: {}, itemFilters: {} }

    const props = {
        allowedFilters: ['Moomin'],
        restrictFilters: true,
    }

    render(
        <Provider store={mockStore(store)}>
            <FilterSelector {...props} />
        </Provider>
    )
    expect(screen.getByRole('button')).toBeTruthy()
})

test('shows button when filters are not restricted', () => {
    const store = { activeModalDimension: {}, itemFilters: {} }

    const props = {
        allowedFilters: [],
        restrictFilters: false,
    }

    render(
        <Provider store={mockStore(store)}>
            <FilterSelector {...props} />
        </Provider>
    )
    expect(screen.getByRole('button')).toBeTruthy()
})
