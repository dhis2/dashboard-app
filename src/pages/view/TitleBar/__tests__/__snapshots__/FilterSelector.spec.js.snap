// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is disabled when offline 1`] = `
<div>
  <span
    class="buttonContainer"
  >
    <button
      class="DropdownButton"
      component="[object Object]"
      disabled=""
      icon="[object Object]"
      type="button"
    >
      Add filter
    </button>
  </span>
</div>
`;

exports[`is enabled when online 1`] = `
<div>
  <span
    class="buttonContainer"
  >
    <button
      class="DropdownButton"
      component="[object Object]"
      icon="[object Object]"
      type="button"
    >
      Add filter
    </button>
  </span>
</div>
`;
